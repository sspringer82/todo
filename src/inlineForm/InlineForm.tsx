import React, { useEffect } from 'react';
import { initialTodo, Todo, TodoInput } from '../Todo';
import useForm from '../useForm';
import Button from '../util/button/Button';
import Input from '../util/input/Input';

export type Props = {
  todo?: Todo;
  onSave: (item: TodoInput) => Promise<void>;
  onCancel: () => void;
};

// @todo generic component?
const InlineForm: React.FC<Props> = ({ todo, onSave, onCancel }) => {
  const { handleSubmit, handleChange, item, setItem } = useForm<TodoInput>(
    initialTodo,
    onSave,
  );
  useEffect(() => {
    if (todo) {
      setItem(todo);
    }
  }, [todo, setItem]);

  return (
    <form onSubmit={handleSubmit} data-testid={`inline-form${todo ? `-${todo.id}` : ''}`} className="m-2 flex">
      <Input
        label="Title"
        type="text"
        onChange={handleChange}
        value={item.title}
        name="title"
        data-testid="title-input"
        style={{width: 230}}
      />
      <Button type="submit" data-testid="submit-button" variant="primary" style={{marginLeft: 5}}>
        save
      </Button>
      <Button
        type="button"
        onClick={() => onCancel()}
        data-testid="cancel-button"
        style={{marginLeft: 5}}
      >
        cancel
      </Button>
    </form>
  );
};

export default InlineForm;
