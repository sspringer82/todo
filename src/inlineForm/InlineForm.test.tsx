import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import InlineForm, { Props } from './InlineForm';

describe('InlineForm', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      onSave: jest.fn(),
      onCancel: jest.fn(),
    };
  });

  it("should perform a snapshot test", () => {
    const {container} = render(<InlineForm {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should show an empty form if no todo is handed over', () => {
    const { getByTestId } = render(<InlineForm {...props} />);
    expect(getByTestId('title-input')).toHaveValue('');
  });
  it('should call onSave with a new item if one should be created', async () => {
    await act(async () => {
      render(<InlineForm {...props} />);
    });

    act(() => {
      fireEvent.change(screen.getByTestId('title-input'), {
        target: { value: 'New Todo' },
      });
    });

    await act(async () => {
      fireEvent.click(screen.getByTestId('submit-button'));
    });
    expect(props.onSave).toHaveBeenCalledWith({
      title: 'New Todo',
      done: false,
      comment: '',
    });
  });
  it('should call onCancel if the cancel button is hit', () => {
    act(() => {
      render(<InlineForm {...props} />);
    });

    act(() => {
      fireEvent.click(screen.getByTestId('cancel-button'));
    });
    expect(props.onCancel).toHaveBeenCalledWith();
  });

  it('should clear the form if the cancel button is hit', () => {
    act(() => {
      render(<InlineForm {...props} />);
    });

    act(() => {
      fireEvent.change(screen.getByTestId('title-input'), {
        target: { value: 'New Todo' },
      });
    });
    expect(screen.getByTestId('title-input')).toHaveValue('New Todo');

    act(() => {
      fireEvent.click(screen.getByTestId('cancel-button'));
    });
    expect(screen.getByTestId('title-input')).toHaveValue('');
  });
  
  it('should call onSave with a changed todo on edit', async () => {
    props.todo = {
      id: 1,
      title: 'New Todo',
      done: true,
      comment: '',
    };

    await act(async () => {
      render(<InlineForm {...props} />);
    });

    act(() => {
      fireEvent.change(screen.getByTestId('title-input'), {
        target: { value: 'Changed Todo' },
      });
    });

    await act(async () => {
      fireEvent.click(screen.getByTestId('submit-button'));
    });
    expect(props.onSave).toHaveBeenCalledWith({
      id: 1,
      title: 'Changed Todo',
      done: true,
      comment: '',
    });
  });
});
