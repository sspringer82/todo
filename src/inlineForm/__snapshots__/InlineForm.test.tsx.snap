// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineForm should perform a snapshot test 1`] = `
<div>
  <form
    class="flex"
    data-testid="inline-form"
  >
    <div
      class="border relative p-1"
      style="border-color: #34d399;"
    >
      <label
        class="absolute bg-gray-800 text-xs px-1"
        style="top: -12px; left: 0px; color: rgb(52, 211, 153);"
      >
        Title
      </label>
      <input
        class="bg-gray-900 py-1 pl-1 focus:outline-none"
        data-testid="title-input"
        label="Title"
        name="title"
        style="width: 230px;"
        type="text"
        value=""
      />
    </div>
    <button
      class="px-5 py-1 bg-opacity-25 focus:outline-none border"
      data-testid="submit-button"
      style="border-color: #34d399; background-color: rgba(52, 211, 153, 0.25); margin-left: 5px;"
      type="submit"
      variant="primary"
    >
      save
    </button>
    <button
      class="px-5 py-1 bg-opacity-25 focus:outline-none border"
      data-testid="cancel-button"
      style="border-color: #34d399; margin-left: 5px;"
      type="button"
    >
      cancel
    </button>
  </form>
</div>
`;
